Add 1.27 feature blog article about StatefulSet Start Ordinals (KEP-3335) (#37418)
kubernetes/website: content/en/blog/_posts/2022-12-16-statefulset-migration.md

in a destination cluster, while maintaining application availability. This
enables you to retain *at most one* semantics and
[Rolling Update](/docs/tutorials/stateful-application/basic-stateful-set/#rolling-update)
behavior when orchestrating a migration across clusters.

### Why would I want to use this feature?

Say you're running your StatefulSet in one cluster, and need to migrate it out
to a different cluster. There are many reasons why you would need to do this:
 * **Scalability**: Your StatefulSet has scaled too large for your cluster, and
   has started to disrupt the quality of service for other workloads in your
   cluster.
 * **Isolation**: You're running a StatefulSet in a cluster that is accessed 
   by multiple users, and namespace isolation isn't sufficient.
 * **Cluster Configuration**: You want to move your StatefulSet to a different
   cluster to use some environment that is not available on your current
   cluster.
 * **Control Plane Upgrades**: You want to move your StatefulSet to a cluster
   running an upgraded control plane, and can't handle the risk or downtime of
   in-place control plane upgrades.

> 

Reviewer: Maybe add something around the lack of PVC orchestration with this feature?

Author: Added this commentary to the "What's Next" section.

7136e37d8864db983f334b85d2b094be47c830e5 10.104.0.15:6379@16379 slave 2cff613d763b22c180cd40668da8e452edef3fc8 0 1669764412595 2 connected                                                                                                                    
a8765caed08f3e185cef22bd09edf409dc2bcc61 10.104.0.19:6379@16379 master - 0 1669764411592 1 connected 0-5460                                                                                                                                                   
2cff613d763b22c180cd40668da8e452edef3fc8 10.104.0.17:6379@16379 master - 0 1669764410000 2 connected 5461-10922
```

5. On `destination`, deploy Redis with zero replicas.

```
helm install redis --namespace kep-3335 \
  bitnami/redis-cluster \
  --set persistence.size=1Gi \
  --set cluster.nodes=0 \
  --set redis.extraEnvVars\[0\].name=REDIS_NODES,redis.extraEnvVars\[0\].value="redis-redis-cluster-headless.kep-3335.svc.cluster.local" \
  --set existingSecret=redis-redis-cluster
```

6. Scale down replica `redis-redis-cluster-5` in the source cluster.

```
kubectl patch sts redis-redis-cluster -p '{"spec": {"replicas": 5}}'
```

> 7. Migrate dependencies from `source` to `destination`.

Reviewer 2: You might want to summarize conceptually what these commands are doing to the PVCs/PVs.

Also add a note that this only works if the storage system you're using is able to support importing existing PVs into a new cluster by copying the existing PV. Some storage systems also keep additional metadata about its volumes and they may have a more specialized sequence to import a volume into a cluster.

Author: Good idea, summarized and added a note about potential storage system limitations.

```

9. On the source cluster, check the replication status.

```
kubectl exec -it redis-redis-cluster-0 -- /bin/bash -c \
  "redis-cli -c -h redis-redis-cluster -a $(kubectl get secret redis-redis-cluster -o jsonpath="{.data.redis-password}" | base64 -d) CLUSTER NODES;"
```

You should see that the new replica's address has joined the Redis cluster.

```
2cff613d763b22c180cd40668da8e452edef3fc8 10.104.0.17:6379@16379 myself,master - 0 1669766684000 2 connected 5461-10922
7136e37d8864db983f334b85d2b094be47c830e5 10.108.0.22:6379@16379 slave 2cff613d763b22c180cd40668da8e452edef3fc8 0 1669766685609 2 connected
2ce30362c188aabc06f3eee5d92892d95b1da5c3 10.104.0.14:6379@16379 master - 0 1669766684000 3 connected 10923-16383
961f35e37c4eea507cfe12f96e3bfd694b9c21d4 10.104.0.18:6379@16379 slave a8765caed08f3e185cef22bd09edf409dc2bcc61 0 1669766683600 1 connected
a8765caed08f3e185cef22bd09edf409dc2bcc61 10.104.0.19:6379@16379 master - 0 1669766685000 1 connected 0-5460
7743661f60b6b17b5c71d083260419588b4f2451 10.104.0.16:6379@16379 slave 2ce30362c188aabc06f3eee5d92892d95b1da5c3 0 1669766686613 3 connected

> ```

Reviewer 2: Any conclusion/summary/next steps?

Author: Added a "What's Next" section.

used.

## Background

StatefulSets ordinals provide sequential identities for pod replicas. When using
[`OrderedReady` Pod management](/docs/tutorials/stateful-application/basic-stateful-set/#orderedready-pod-management)
Pods are created from ordinal index `0` up to `N-1`.

With Kubernetes today, orchestrating a StatefulSet migration across clusters is
challenging. Backup and restore solutions exist, but these require the
application to be scaled down to zero replicas prior to migration. In today's
fully connected world, even planned application downtime may not allow you to
meet your business goals. You could use
[Cascading Delete](/docs/tutorials/stateful-application/basic-stateful-set/#cascading-delete)
or
[On Delete](/docs/tutorials/stateful-application/basic-stateful-set/#on-delete)
to migrate individual pods, however this is error prone and tedious to manage.
You lose the self-healing benefit of the StatefulSet controller when your Pods
fail or are evicted.

Kubernetes v1.26 enables a StatefulSet to be responsible for a range of ordinals
within a half-open interval `[0, N)` (the ordinals 0, 1, ... N-1).
With it, you can scale down a range

> (`[0, k)`) in a source cluster, and scale up the complementary range (`[k, N)`)

Author: Updated to use {0..k} ranges. These are used in other StatefulSet documentation (eg: https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/)

in a destination cluster, while maintaining application availability. This

> enables you to retain *at most one* semantics and

Author: Added definition of "at most one" semantics.

behavior when orchestrating a migration across clusters.

## Why would I want to use this feature?

Say you're running your StatefulSet in one cluster, and need to migrate it out
to a different cluster. There are many reasons why you would need to do this:
 * **Scalability**: Your StatefulSet has scaled too large for your cluster, and
   has started to disrupt the quality of service for other workloads in your
   cluster.
 * **Isolation**: You're running a StatefulSet in a cluster that is accessed 
   by multiple users, and namespace isolation isn't sufficient.
 * **Cluster Configuration**: You want to move your StatefulSet to a different
   cluster to use some environment that is not available on your current
   cluster.
 * **Control Plane Upgrades**: You want to move your StatefulSet to a cluster
   running an upgraded control plane, and can't handle the risk or downtime of
   in-place control plane upgrades.

## How do I use it?

Enable the `StatefulSetStartOrdinal` feature gate on a cluster, and create a
StatefulSet with a customized `.spec.ordinals.start`.

## Try it for yourself

In this demo, you'll use the `StatefulSetStartOrdinal` feature to migrate a

> StatefulSet from one Kubernetes cluster to another. For this demo, the

Author: Updated to first person singular, here and below.

[redis-cluster](https://github.com/bitnami/charts/tree/main/bitnami/redis-cluster)
Bitnami Helm chart is used to install Redis.

Tools Required:
 * [yq](https://github.com/mikefarah/yq)
 * [helm](https://helm.sh/docs/helm/helm_install/)

> Pre-requisites: Two Kubernetes clusters named `source` and `destination`.
>  * `StatefulSetStartOrdinal` feature gate is enabled on both clusters
>  * The same default `StorageClass` is installed on both clusters. This
>    `StorageClass` should provision underlying storage that is accessible from
>    both clusters.
>  * A flat network topology that allows for pods to be accessible across both
>    Kubernetes clusters. If creating clusters on a cloud provider, this
>    configuration may be called private cloud or private network.

Reviewer 3: How about:
```suggestion
### Pre-requisites {#demo-pre-requisites}

To do this, I need two Kubernetes clusters that can both access common networking and
storage; I've named my clusters `source` and `destination`.
Specifically, I need:

* The `StatefulSetStartOrdinal` feature gate enabled on both clusters.
* Client configuration for `kubectl` that lets me access both clusters as an administrator.
* The same `StorageClass` installed on both clusters, and set as the default
  StorageClass for both clusters. This `StorageClass` should provision underlying
  storage that is accessible from either or both clusters.
* A flat network topology that allows for pods to send and receive packets to and
  from Pods in either clusters. If you are creating clusters on a cloud provider, this
  configuration may be called private cloud or private network.
```

Author: Thanks for the suggestion, it's more readable and descriptive.

1. Create a demo namespace on both clusters.

   ```
   kubectl create ns kep-3335
   ```

> 2. Deploy a Redis cluster on `source`.

Reviewer 3: ```suggestion
2. Deploy a Redis cluster in the `source` cluster:
```

Author: Updated

   ```
   helm repo add bitnami https://charts.bitnami.com/bitnami
   helm install redis --namespace kep-3335 \
     bitnami/redis-cluster \

>      --set persistence.size=1Gi

Reviewer 3: To help illustrate what's happening, I'd be explicit about the number of nodes that get created in `source`:
```suggestion
     --set persistence.size=1Gi \
     --set cluster.nodes=6
```

Author: Good call, added the `--cluster.nodes=6` param

   ```

3. On `source`, check the replication status.

   ```
   kubectl exec -it redis-redis-cluster-0 -- /bin/bash -c \

>      "redis-cli -c -h redis-redis-cluster -a $(kubectl get secret redis-redis-cluster -o jsonpath="{.data.redis-password}" | base64 -d) CLUSTER NODES;"

Reviewer 3: Is this more useful?
```suggestion
     "redis-cli -c -h redis-redis-cluster -a $(kubectl get secret redis-redis-cluster -o jsonpath="{.data.redis-password}" | base64 -d) 'CLUSTER NODES;'"
```
I don't know `redis-cli` to tell if it is.

Author: `redis-cli` does require `CLUSTER` and `NODES` to be separate command line arguments, so leaving as is.

   7743661f60b6b17b5c71d083260419588b4f2451 10.104.0.16:6379@16379 slave 2ce30362c188aabc06f3eee5d92892d95b1da5c3 0 1669764410000 3 connected                                                                                             
   961f35e37c4eea507cfe12f96e3bfd694b9c21d4 10.104.0.18:6379@16379 slave a8765caed08f3e185cef22bd09edf409dc2bcc61 0 1669764411000 1 connected                                                                                                             
   7136e37d8864db983f334b85d2b094be47c830e5 10.104.0.15:6379@16379 slave 2cff613d763b22c180cd40668da8e452edef3fc8 0 1669764412595 2 connected                                                                                                                    
   a8765caed08f3e185cef22bd09edf409dc2bcc61 10.104.0.19:6379@16379 master - 0 1669764411592 1 connected 0-5460                                                                                                                                                   
   2cff613d763b22c180cd40668da8e452edef3fc8 10.104.0.17:6379@16379 master - 0 1669764410000 2 connected 5461-10922
   ```

> 4. On `destination`, deploy Redis with zero replicas.

Reviewer 3: ```suggestion
4. On `destination`, deploy Redis with zero replicas:
```

Author: Updated these ordered list bullets to start with action verbs, here and elsewhere

   ```
   helm install redis --namespace kep-3335 \
     bitnami/redis-cluster \
     --set persistence.size=1Gi \
     --set cluster.nodes=0 \
     --set redis.extraEnvVars\[0\].name=REDIS_NODES,redis.extraEnvVars\[0\].value="redis-redis-cluster-headless.kep-3335.svc.cluster.local" \
     --set existingSecret=redis-redis-cluster
   ```

> 5. Scale down replica `redis-redis-cluster-5` in the source cluster.

Reviewer 3: ```suggestion
5. Scale down Redis in the source cluster to remove the sixth Pod, `redis-redis-cluster-5`:
```

Author: Updated to match phrasing from bullet 7

   ```
   kubectl patch sts redis-redis-cluster -p '{"spec": {"replicas": 5}}'
   ```

6. Migrate dependencies from `source` to `destination`.

   The following commands copy resources from `source` to `destionation`. Details
   that are not relevant in `destination` cluster are removed (eg: `uid`,
   `resourceVersion`, `status`).

>    #### Source cluster

Reviewer 3: Don't put headings inside ordered lists.
```suggestion
   **Steps for the source cluster**
```

Author: Updated

   Note: If using a `StorageClass` with `reclaimPolicy: Delete` configured, you
         should patch the PVs in `source` with `reclaimPolicy: Retain` prior to
         deletion to retain the underlying storage used in `destination`. See
         [Change the Reclaim Policy of a PersistentVolume](/docs/tasks/administer-cluster/change-pv-reclaim-policy/)
         for more details.

   ```
   kubectl get pvc redis-data-redis-redis-cluster-5 -o yaml | yq 'del(.metadata.uid, .metadata.resourceVersion, .metadata.annotations, .metadata.finalizers, .status)' > /tmp/pvc-redis-data-redis-redis-cluster-5.yaml
   kubectl get pv $(yq '.spec.volumeName' /tmp/pvc-redis-data-redis-redis-cluster-5.yaml) -o yaml | yq 'del(.metadata.uid, .metadata.resourceVersion, .metadata.annotations, .metadata.finalizers, .spec.claimRef, .status)' > /tmp/pv-redis-data-redis-redis-cluster-5.yaml
   kubectl get secret redis-redis-cluster -o yaml | yq 'del(.metadata.uid, .metadata.resourceVersion)' > /tmp/secret-redis-redis-cluster.yaml
   ```

>    #### Destination cluster

Reviewer 3: Don't put headings inside ordered lists.
```suggestion
   **Steps for the destination cluster**
```

Author: Updated

   Note: For the PV/PVC, this procedure only works if the underlying storage system
         that your PVs use can support being copied into `destination`. Storage
         that is associated with a specific node or topology may not be supported.
         Additionally, some storage systems may store addtional metadata about
         volumes outside of a PV object, and may require a more specialized
         sequence to import a volume.

   ```
   kubectl create -f /tmp/pv-redis-data-redis-redis-cluster-5.yaml
   kubectl create -f /tmp/pvc-redis-data-redis-redis-cluster-5.yaml
   kubectl create -f /tmp/secret-redis-redis-cluster.yaml
   ```

> 7. Scale up replica `redis-redis-cluster-5` in the destination cluster, with a
>    start ordinal of 5:

Reviewer 3: ```suggestion
7. Scale up the `redis-redis-cluster` StatefulSet in the destination cluster by 1, with a
   start ordinal of 5:
```

Author: Updated

   ```
   kubectl patch sts redis-redis-cluster -p '{"spec": {"ordinals": {"start": 5}, "replicas": 1}}'
   ```

> 8. On the source cluster, check the replication status.

Reviewer 3: ```suggestion
8. In the source cluster, check the Redis replication status:
```

Author: Update to action verb phrase

   ```
   kubectl exec -it redis-redis-cluster-0 -- /bin/bash -c \
     "redis-cli -c -h redis-redis-cluster -a $(kubectl get secret redis-redis-cluster -o jsonpath="{.data.redis-password}" | base64 -d) CLUSTER NODES;"
   ```

>    You should see that the new replica's address has joined the Redis cluster.

Reviewer 3: ```suggestion
   I should see that the new replica has joined the Redis cluster (look for its IP address).
```

   ```
   2cff613d763b22c180cd40668da8e452edef3fc8 10.104.0.17:6379@16379 myself,master - 0 1669766684000 2 connected 5461-10922
   7136e37d8864db983f334b85d2b094be47c830e5 10.108.0.22:6379@16379 slave 2cff613d763b22c180cd40668da8e452edef3fc8 0 1669766685609 2 connected
   2ce30362c188aabc06f3eee5d92892d95b1da5c3 10.104.0.14:6379@16379 master - 0 1669766684000 3 connected 10923-16383
   961f35e37c4eea507cfe12f96e3bfd694b9c21d4 10.104.0.18:6379@16379 slave a8765caed08f3e185cef22bd09edf409dc2bcc61 0 1669766683600 1 connected
   a8765caed08f3e185cef22bd09edf409dc2bcc61 10.104.0.19:6379@16379 master - 0 1669766685000 1 connected 0-5460
   7743661f60b6b17b5c71d083260419588b4f2451 10.104.0.16:6379@16379 slave 2ce30362c188aabc06f3eee5d92892d95b1da5c3 0 1669766686613 3 connected
   ```

> 

Reviewer 3: I'd expect to see:
```markdown
9. Scale down Redis in the source cluster to remove the fifth Pod, `redis-redis-cluster-4`:
```
and then something like
```markdown
10. Scale up the `redis-redis-cluster` StatefulSet in the destination cluster by 1, with a
   start ordinal of FIXME:

**THINGS HERE**

11. Repeat this until done.

I continue until the StatefulSet for Redis in the source cluster is scaled to 0, and Redis
in the destination cluster is healthy with 6 total nodes.
```

Author: Added a 9th step to instruct to the reader to iterate through steps 5-7

## What's Next?

This feature provides a building block for a StatefulSet to be split up across
clusters, but does not prescribe the mechanism as to how the StatefulSet should
be migrated. Migration requires coordination of StatefulSet replicas, along with
orchestration of the storage and network layer. This is dependent on the storage
and connectivity requirements of the application installed by the StatefulSet.

> Additionally, many StatefulSets are controlled by Operators, which adds another

Reviewer 3: Optionally:
```suggestion
Additionally, many StatefulSets are managed by
[operators](/docs/concepts/extend-kubernetes/operator/), which adds another
```

layer of complexity to migration.

> If you're interested in building blocks to make these processes easier, get

Reviewer 3: Optionally:
```suggestion
If you're interested in building better blocks to make these processes easier, get
```

Author: Updated to "enhancements"

  --set existingSecret=redis-redis-cluster
```

6. Scale down replica `redis-redis-cluster-5` in the source cluster.

```
kubectl patch sts redis-redis-cluster -p '{"spec": {"replicas": 5}}'
```

7. Migrate dependencies from `source` to `destination`.

The following commands copy resources from `source` to `destionation`. Details
that are not relevant in `destination` cluster are removed (eg: `uid`,
`resourceVersion`, `status`).

Note: For the PV/PVC, this procedure only works if the underlying storage system
      that your `StorageClass` uses can support copying existing PVs into a
      new cluster. Storage that is associated with a specific node or topology
      may not be supported. Additionally, some storage systems may store
      addtional metadata about volumes outside of a PV object, and may require
      a more specialized sequence to import a volume.

Source Cluster

```
kubectl get pvc redis-data-redis-redis-cluster-5 -o yaml | yq 'del(.metadata.uid, .metadata.resourceVersion, .metadata.annotations, .metadata.finalizers, .status)' > /tmp/pvc-redis-data-redis-redis-cluster-5.yaml

> kubectl get pv $(yq '.spec.volumeName' /tmp/pvc-redis-data-redis-redis-cluster-5.yaml) -o yaml | yq 'del(.metadata.uid, .metadata.resourceVersion, .metadata.annotations, .metadata.finalizers, .spec.claimRef, .status)' > /tmp/pv-redis-data-redis-redis-cluster-5.yaml

Reviewer 2: Should we set the PV reclaimPolicy to retain to avoid accidental disk deletion in the source cluster?

Author: Good idea, added a note to a reference guide on how to do so.

---
layout: blog

> title: "Kubernetes 1.26: StatefulSet Migration"

Reviewer 3: If we do keep the MCS aspect of this article, then maybe this:
```suggestion
title: "Kubernetes 1.26: StatefulSet Numbering Helps With Stateful Workload Migration"
```

The post-release articles follow a pattern: they introduce a new feature and explain how it's useful. We shouldn't omit the first bit, the introduction. Imagine you're at a SIG Apps meeting and you're explaining how this new feature, not yet implemented, will benefit end users. What would you say?

Author: Updated the title to reflect the benefit of this feature, and goal of the tutorial

> date: 2022-12-16

Reviewer 3: ```suggestion
date: 2023-01-03
```

Author: Updated

slug: statefulset-migration
---

**Author**: Peter Schuurman (Google)

Kubernetes v1.26 introduces a new, alpha-level feature for
[StatefulSets](/docs/concepts/workloads/controllers/statefulset/) that controls
the ordinal numbering of Pod replicas. Ordinals can start from arbitrary
non-negative numbers. This blog post will discuss how this feature can be
used.

## Background

StatefulSets ordinals provide sequential identities for pod replicas. When using

> [OrderedReady Pod Management](/docs/tutorials/stateful-application/basic-stateful-set/#orderedready-pod-management),

Reviewer 3: ```suggestion
[`OrderedReady` Pod management](/docs/tutorials/stateful-application/basic-stateful-set/#orderedready-pod-management)
for a StatefulSet that has _N_ replicas,
```

Pods are created from ordinal index `0` up to `N-1`.

With Kubernetes today, orchestrating a StatefulSet migration across clusters is
challenging. Backup and restore solutions exist, but these require the
application to be scaled down to zero replicas prior to migration. In today's

> fully connected world, planned downtime and unavailability may not allow you to

Reviewer 3: ```suggestion
fully connected world, even planned application downtime may not allow you to
```

meet your business goals. You could use
[Cascading Delete](/docs/tutorials/stateful-application/basic-stateful-set/#cascading-delete)
or
[On Delete](/docs/tutorials/stateful-application/basic-stateful-set/#on-delete)
to migrate individual pods, however this is error prone and tedious to manage.
You lose the self-healing benefit of the StatefulSet controller when your Pods
fail or are evicted.

> This feature enables a StatefulSet to be responsible for a range of ordinals

Reviewer 3: ```suggestion
Kubernetes v1.26 enables a StatefulSet to be responsible for a range of ordinals
```

within a logical range of `[0, N)`. With it, you can scale down a range

> (`[0, k)`) in a source cluster, and scale up the complementary range (`[k, N)`)

Reviewer 3: You might need to explain the notation here for readers who are not mathematicians.

Author: Updated with an example, and a reference to half-open interval

behavior when orchestrating a migration across clusters.

## Why would I want to use this feature?

Say you're running your StatefulSet in one cluster, and need to migrate it out
to a different cluster. There are many reasons why you would need to do this:
 * **Scalability**: Your StatefulSet has scaled too large for your cluster, and
   has started to disrupt the quality of service for other workloads in your
   cluster.
 * **Isolation**: You're running a StatefulSet in a cluster that is accessed 
   by multiple users, and namespace isolation isn't sufficient.
 * **Cluster Configuration**: You want to move your StatefulSet to a different
   cluster to use some environment that is not available on your current
   cluster.
 * **Control Plane Upgrades**: You want to move your StatefulSet to a cluster
   running an upgraded control plane, and can't handle the risk or downtime of
   in-place control plane upgrades.

## How do I use it?

Enable the `StatefulSetStartOrdinal` feature gate on a cluster, and create a
StatefulSet with a customized `.spec.ordinals.start`.

## Try it for yourself

In this demo, you'll use the `StatefulSetStartOrdinal` feature to migrate a

> StatefulSet from one cluster to another. For this demo, the

Reviewer 3: ```suggestion
StatefulSet from one Kubernetes cluster to another. For this demo, the
```

[redis-cluster](https://github.com/bitnami/charts/tree/main/bitnami/redis-cluster)
Bitnami Helm chart is used to install Redis.

Tools Required:
 * [yq](https://github.com/mikefarah/yq)
 * [helm](https://helm.sh/docs/helm/helm_install/)

Pre-requisites: Two clusters named `source` and `destination`.
 * `StatefulSetStartOrdinal` feature gate is enabled on both clusters
 * [MultiClusterServices](https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api)

> support is enabled

Reviewer 3: :warning: We shouldn't write a blog article that relies on an undocumented feature / API

Unfortunately, MCS isn't part of the Kubernetes documentation and nor is their in-project documentation elsewhere, that I know of.

Examples that would be OK to link to:
- https://gateway-api.sigs.k8s.io/
- https://kubernetes-csi.github.io/docs/

Author: Removed the reference for MCS for now. Once this is documented upstream, I'll add it back in. The nice thing about Redis is that migration works with a flat network topology, irrespective of the services available, due to caching of existing placements. This makes MCS unnecessary for Redis if the clusters are created on the same flat network. However having a documented tutorial with MCS would be helpful for a user wanting to migrate another workload that doesn't have this capability.

 * The same default `StorageClass` is installed on both clusters. This
   `StorageClass` should provision underlying storage that is accessible from

>    both clusters

Reviewer 3: ```suggestion
   both clusters.
```

> 1. Create a demo namespace on both clusters.

Reviewer 3: If using an ordered list (this _is_ an ordered list according to CommonMark), then indent the text up to the start of the next list item, so that it matches the first character of the text.

Avoid 4-character indents because in Markdown a 4 character indent can also mean monospace block.

Author: Thanks, updated these to 3-indent to match the first character of the text in the ordered list.

a8765caed08f3e185cef22bd09edf409dc2bcc61 10.104.0.19:6379@16379 master - 0 1669764411592 1 connected 0-5460                                                                                                                                                   
2cff613d763b22c180cd40668da8e452edef3fc8 10.104.0.17:6379@16379 master - 0 1669764410000 2 connected 5461-10922
```

5. On `destination`, deploy Redis with zero replicas.

```
helm install redis --namespace kep-3335 \
  bitnami/redis-cluster \
  --set persistence.size=1Gi \
  --set cluster.nodes=0 \
  --set redis.extraEnvVars\[0\].name=REDIS_NODES,redis.extraEnvVars\[0\].value="redis-redis-cluster-headless.kep-3335.svc.cluster.local" \
  --set existingSecret=redis-redis-cluster
```

6. Scale down replica `redis-redis-cluster-5` in the source cluster.

```
kubectl patch sts redis-redis-cluster -p '{"spec": {"replicas": 5}}'
```

7. Migrate dependencies from `source` to `destination`.

The following commands copy resources from `source` to `destionation`. Details
that are not relevant in `destination` cluster are removed (eg: `uid`,
`resourceVersion`, `status`).

> Source Cluster

Reviewer 3: ```suggestion
#### Source cluster
```

Note: If using a `StorageClass` with `reclaimPolicy: Delete` configured, you
      should patch the PVs in `source` with `reclaimPolicy: Retain` prior to
      deletion to retain the underlying storage used in `destination`. See
      [Change the Reclaim Policy of a PersistentVolume](/docs/tasks/administer-cluster/change-pv-reclaim-policy/)
      for more details.

```
kubectl get pvc redis-data-redis-redis-cluster-5 -o yaml | yq 'del(.metadata.uid, .metadata.resourceVersion, .metadata.annotations, .metadata.finalizers, .status)' > /tmp/pvc-redis-data-redis-redis-cluster-5.yaml
kubectl get pv $(yq '.spec.volumeName' /tmp/pvc-redis-data-redis-redis-cluster-5.yaml) -o yaml | yq 'del(.metadata.uid, .metadata.resourceVersion, .metadata.annotations, .metadata.finalizers, .spec.claimRef, .status)' > /tmp/pv-redis-data-redis-redis-cluster-5.yaml
kubectl get secret redis-redis-cluster -o yaml | yq 'del(.metadata.uid, .metadata.resourceVersion)' > /tmp/secret-redis-redis-cluster.yaml
```

> Destination Cluster

Reviewer 3: ```suggestion
#### Destination cluster
```

Note: For the PV/PVC, this procedure only works if the underlying storage system
      that your PVs use can support being copied into `destination`. Storage
      that is associated with a specific node or topology may not be supported.
      Additionally, some storage systems may store addtional metadata about
      volumes outside of a PV object, and may require a more specialized
      sequence to import a volume.

```
kubectl create -f /tmp/pv-redis-data-redis-redis-cluster-5.yaml
kubectl create -f /tmp/pvc-redis-data-redis-redis-cluster-5.yaml
kubectl create -f /tmp/secret-redis-redis-cluster.yaml
```

> 8. Scale up replica `redis-redis-cluster-5` in the destination cluster.

Reviewer 3: ```suggestion
8. Scale up replica `redis-redis-cluster-5` in the destination cluster, starting numbering from 5:
```

---
layout: blog
title: "Kubernetes 1.26: StatefulSet Start Ordinal Simplifies Migration"

> date: 2023-04-19

Reviewer 4: ```suggestion
date: 2023-04-28
```
Pls also change the file path to match the new date.

Author: Thanks! Updated.
